<center><h1><strong><font color="blue">Studi Kasus Data Wrangling</font></strong></h1></center>

![taudata Analytics](https://github.com/taudataid/eLearning/raw/master/images/covers/cover_ui-n-taudata.jpg "taudata Analytics")

## Group D
Ketua : Muhammad Hanif -  Zamzami - 2006487566

Anggota :
*   Faza Ilmiyah - 2106638620
*   Hosia Josindra - 2006463332
*   Laras Kirana - 2106653035
*   Marcella Sintauly - 2006568916
*   Nadia Amelia - 2106653224



In [ ]:
import warnings; warnings.simplefilter('ignore')
import pandas as pd

try: # Running Locally, yakinkan "file_" berada di folder "data"
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')
except: # Running in Google Colab
    !mkdir data
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_nilai.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/mengajar.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_dosen.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_kelas.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_krs.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mahasiswa.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mata_kuliah.csv
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')

mengajar.shape, dosen.shape, kelas.shape, krs.shape, mahasiswa.shape, matkul.shape, nilai.shape

--2022-11-18 02:16:27--  https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_nilai.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1136 (1.1K) [text/plain]
Saving to: ‘data/tb_nilai.csv’

tb_nilai.csv        100%[===================>]   1.11K  --.-KB/s    in 0s      

2022-11-18 02:16:27 (43.4 MB/s) - ‘data/tb_nilai.csv’ saved [1136/1136]

--2022-11-18 02:16:27--  https://raw.githubusercontent.com/taudataid/eLearning/master/data/mengajar.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 329 [text/plain]
Savi

((14, 4), (4, 6), (4, 4), (7, 6), (15, 8), (9, 3), (33, 7))

# Code 01

In [ ]:
# Code untuk Soal No 01 diberikan sebagai Contoh dan Bonus
# Pelajari dengan baik

nilai_dosen = nilai.groupby(nilai['NIDN']).mean()[['nilai']]
nilai_dosen = nilai_dosen.rename({'nilai': 'mean_nilai'}, axis=1)
nilai_dosen = pd.merge(dosen, nilai_dosen, left_on='NIDN', right_index=True)
nilai_dosen = nilai_dosen[['NIDN', 'nama', 'mean_nilai']]
sb = nilai.groupby(nilai['NIDN']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1)
nilai_dosen = nilai_dosen.set_index(['NIDN'])
nilai_dosen = pd.concat([nilai_dosen, sb], axis=1)
nilai_dosen

nama  mean_nilai  std_nilai
NIDN                                      
123    Andri Basuki   85.750000   7.741355
234        zulfikar   80.428571  15.883504
345   Annisa Fitria   60.090909  16.220077
787            yudi   70.571429  13.010984

# Interpretasi 01:

* Pada data terlihat bahwa terdapat perbedaan rata-rata nilai yang diberikan setiap dosen. Andri Basuki memiliki rata-rata nilai paling tinggi, yaitu 85.750000. Pada urutan ke-2 adalah Zulfikar dengan rata-rata nilai, yaitu 80.428571. Pada urutan ke-3 adalah Yudi dengan rata-rata nilai, yaitu 70.571429. Pada urutan ke-4 adalah Annisa Fitria dengan rata-rata nilai, yaitu 60.090909.

* Rata-rata nilai Andri Basuki memiliki std terkecil (7.741355), artinya lebih mendekati nilai rata-rata dan variansi datanya tidak tersebar.

* Rata-rata nilai Annisa Fitria memiliki std terbesar (16.220077), artinya variansi datanya akan lebih tersebar.

kesimpulannya, terdapat perbedaan pemberian nilai antar dosen. Terdapat dosen yang cenderung memberikan nilai cenderung lebih tinggi (atau lebih rendah) dibandingkan dosen lainnya.

# Code 02

In [ ]:
# Code Untuk Soal No 02

nd_df = pd.merge(dosen, nilai, on='NIDN')
mean = nd_df[['Pendidikan_Terakhir', 'nilai' ]]
mean = mean.groupby(nd_df['Pendidikan_Terakhir']).mean()[['nilai']]
mean = mean.rename({'nilai': 'mean_nilai'}, axis=1)

std = nd_df.groupby(nd_df['Pendidikan_Terakhir']).std()[['nilai']]
std = std.rename({'nilai': 'std_nilai'}, axis=1)

nilaipt_dosen = pd.concat([mean, std], axis=1)
nilaipt_dosen

mean_nilai  std_nilai
Pendidikan_Terakhir                       
S2                    64.166667  15.561075
S3                    83.266667  12.068061

# Interpretasi 02:

* Terlihat pada data bahwa rata - rata pemberian nilai dosen dengan pendidikan terakhir S3 lebih tinggi daripada dosen dengan pendidikan terakhir S2. Rata-rata pemberian nilai dosen dengan pendidikan terakhir S3 sebesar 83.266667, sedangkan rata-rata pemberian nilai dosen dengan pendidikan terakhir S2 sebesar 64.166667. Pengaruh std_nilai terhadap mean_nilai adalah semakin rendah nilai std, semakin dekat mendekati rata-rata (tidak tersebar). Sementara itu, jika nilai std tinggi, akan semakin lebar rentang variasi datanya (semakin tersebar).

- Rata-rata pemberian nilai dosen dengan pendidikan terakhir S3, lebih mendekati nilai rata-rata dengan std yang kecil (12)
- Rata-rata pemberian nilai dosen dengan pendidikan terakhir S2, memiliki variasi data yang tersebar dengan std yang tinggi (15)
- Besar nilai standar deviasi merepresentasikan pesebaran data. Semakin besar nilai std-nya maka semakin beragam nilainya atau semakin tidak akurat nilai mean nya

Kesimpulannya, terdapat perbedaan yang signifikan dalam pemberian nilai antara dosen dengan pendidikan terakhir S3 dan S2.





# Code 03

In [ ]:
# Code Soal No 03

nilai_m = nilai[['nim','nilai']]
mahasiswa_m = mahasiswa[['nim', 'Asal_Sekolah']]
int_3 = pd.merge(mahasiswa_m, nilai_m, on = 'nim', how = 'left')
mean = int_3.groupby('Asal_Sekolah')[['nilai']].agg('mean')
mean = mean.rename({'nilai': 'mean_nilai'}, axis=1)

std = int_3.groupby(int_3['Asal_Sekolah']).std()[['nilai']]
std = std.rename({'nilai': 'std_nilai'}, axis=1)

pd.concat([mean,std],axis = 1)

mean_nilai  std_nilai
Asal_Sekolah                       
MA             70.777778  21.111477
SMK            82.818182  11.771306
SMU            65.846154  14.200036

# Interpretasi 03:

* Terlihat pada tabel, mahasiswa yang pendidikan terakhirnya berasal dari SMK memiliki rata-rata nilai tertinggi selama perkuliahan, yaitu 82.81. Pada urutan ke-2 tertinggi adalah mahasiswa yang pendidikan terakhirnya berasal dari MA (Madrasah) dan urutan terakhir adalah mahasiswa dengan pendidikan terakhir dari SMU dengan nilai 65.84.

- Rata-rata nilai dari mahasiswa berasal dari SMK, memiliki std terkecil (11.77) yang artinya lebih dekat mendekati nilai rata-rata dan variasi datanya tidak tersebar.
- Rata-rata nilai dari mahasiswa berasal dari MA, memiliki std terbesar (21.11) yang artinya variasi datanya akan semakin terbsebar.

Kesimpulanya, performa akademik yang dilihat dari rata-rata nilai selama perkuliahan memiliki perbedaan yang cukup signifikan berdasarkan pendidikan terakhir mahasiswa.

# Code 4

In [ ]:
# Code Soal No 04

nilai_m = nilai[['nim','nilai']]
mahasiswa_m = mahasiswa[['nim', 'Gender']]
int_4 = pd.merge(mahasiswa_m, nilai_m, on = 'nim', how = 'left')
mean = int_4.groupby('Gender')[['nilai']].agg('mean')
mean = mean.rename({'nilai': 'mean_nilai'}, axis=1)

std = int_4.groupby(int_4['Gender']).std()[['nilai']]
std = std.rename({'nilai': 'std_nilai'}, axis=1)

pd.concat([mean,std],axis = 1)


mean_nilai  std_nilai
Gender                       
Pria     73.888889  17.146047
Wanita   71.600000  17.112235

# Interpretasi 04

* Berdasarkan rata-rata nilai yang di peroleh, yaitu Pria (mahasiswa) dan Wanita (mahasiswi). Pada tabel, terlihat rata-rata nilai pria/mahasiswa lebih tinggi (73.89) dibanding nilai rata-rata wanita/mahasiswa (71.60)

* Rata-rata nilai dari mahasiswa, memiliki std kecil (17.14) yang artinya lebih dekat mendekati nilai rata-rata dan variasi datanya tidak tersebar.

* Rata-rata nilai dari mahasiswi, memiliki std kecil (17.11) yang artinya variasi datanya lebih dekat mendekati nilai rata-rata dan variasi datanya tidak tersebar.

Dapat disimpulkan terdapat perbedaan kemampuan akademik antara mahasiswa dan mahasiswi jika dilihat dari niai rata-rata.

# Code 05

In [ ]:
# Code soal No 05

nm_df = pd.merge(nilai, mahasiswa, on = 'nim')
nm_df = nm_df[['Angkatan', 'id_mata_kuliah', 'nilai']]

pa_mean = nm_df.groupby(['Angkatan', 'id_mata_kuliah'])[['nilai']].mean()
pa_mean = pa_mean.rename({'nilai': 'mean_nilai'}, axis=1)

pa_std = nm_df.groupby(['Angkatan', 'id_mata_kuliah']).std()[['nilai']]
pa_std = pa_std.rename({'nilai': 'std_nilai'}, axis=1)

pa_angkatan = pd.concat([pa_mean, pa_std], axis=1)
pa_angkatan


mean_nilai  std_nilai
Angkatan id_mata_kuliah                       
2019     CS01             80.500000  17.253019
         KAL01            85.750000   7.741355
         KAL02            63.500000   8.103497
         STA02            54.000000  12.543258
2020     CS01             80.333333  17.616280
         KAL01            70.571429  13.010984
         STA02            63.666667  29.399546

# Interpretasi 05

Prestasi akademik akan dihitung berdasarkan angkatan, mata kuliah, dan rata-rata nilai. Berdasarkan hasil pada tabel, dapat diambil beberapa kesimpulan sebagai berikut:

1. Mata kuliah CS01 angkatan 2019 memiliki index nilai rata-rata yang lebih tinggi dengan nilai 80.50 dibanding dengan CS01 angkatan 2020 sebesar 80.33, walaupun hanya memiliki perbedaan yang sedikit (tidak terlalu signifikan)

2. Mata kuliah KAL01 angkatan 2019 memiliki index nilai rata-rata yang lebih tinggi dengan nilai 85.75 dibanding dengan KAL01 angkatan 2020 dengan nilai 70.57. KAL01 angkatan 2019 memiliki std lebih kecil sehingga akan semakin mendekati rata-rata.

3. Mata kuliah STA02 angkatan 2019 memiliki index nilai rata-rata yang lebih kecil dengan nilai 54.00, dibanding STA02 angkatan 2020 yang memiliki nilai sebesar 63.67. std untuk STA02 angkatan 2019 terlihat tinggi dibanding std untuk STA02 angkatan 2020 sehingga persebaran / variasi datanya semakin tersebar.

Kesimpulan, angkatan 2019 dengan perbandingan 2 dari 3 memiliki index rata-rata nilai lebih tinggi.

# Code 06 #1

In [ ]:
# Code soal No 06 #1
# Permasalahan perbedaan dalam pemberian nilai untuk Dosen Pria dan Wanita

nd_df = pd.merge(dosen, nilai, on='NIDN')
jk_mean = nd_df[['Jenis_Kelamin', 'nilai' ]]
jk_mean = jk_mean.groupby(nd_df['Jenis_Kelamin']).mean()[['nilai']]
jk_mean = jk_mean.rename({'nilai': 'mean_nilai'}, axis=1)

jk_std = nd_df.groupby(nd_df['Jenis_Kelamin']).std()[['nilai']]
jk_std = jk_std.rename({'nilai': 'std_nilai'}, axis=1)

nilaijk_dosen = pd.concat([jk_mean, jk_std], axis=1)
nilaijk_dosen

mean_nilai  std_nilai
Jenis_Kelamin                       
Pria            79.227273  13.494026
Wanita          60.090909  16.220077

# Interpretasi 06 #1
Terdapat informasi lain dalam data, yaitu berdasarkan data wrangling di atas, akan ditinjau kecenderungan pemberian nilai oleh dosen berdasarkan gender

* Dosen wanita memiliki mean pemberian nilai 60.090
* Dosen pria memiliki mean pemberian nilai 79.227
* Dosen wanita cenderung memberikan nilai yang lebih kecil dibandingkan dosen Pria

Dapat disimpulkan bahwa gender dosen berpengaruh pada pemberian nilai

# Code 06 #2

In [ ]:
# Code soal No 06 #2

nilai_m = nilai[['id_mata_kuliah','nilai']]
mean = nilai_m.groupby('id_mata_kuliah')[['nilai']].agg('mean')
mean = mean.rename({'nilai': 'mean_nilai'}, axis=1)

std = nilai_m.groupby(nilai_m['id_mata_kuliah']).std()[['nilai']]
std = std.rename({'nilai': 'std_nilai'}, axis=1)
left = pd.concat([mean,std],axis = 1)

matkul = pd.DataFrame(nilai['id_mata_kuliah'].value_counts())
right = matkul.rename({'id_mata_kuliah': 'frekuensi_mahasiswa_per_kelas'}, axis=1)
right.reset_index(inplace=True)
right = right.rename(columns = {'index':'matkul'})

matakuliah = pd.merge(left, right, left_on = 'id_mata_kuliah', right_on = 'matkul')
matakuliah = matakuliah.set_index('matkul')
matakuliah


mean_nilai  std_nilai  frekuensi_mahasiswa_per_kelas
matkul                                                      
CS01     80.428571  15.883504                              7
KAL01    78.666667  12.804389                             15
KAL02    63.500000   8.103497                              4
STA02    58.142857  19.836234                              7

# Interpretasi 06 #2

* Terdapat informasi lain dalam data, yaitu berdasarkan data wrangling di atas, didapat mata kuliah dengan jumlah mahasiswa di kelas mata kuliah tersebut tidak berpengaruh terhadap rata-rata nilai mahasiswa tsb. Dengan kelas KAL01 yang memiliki jumlah mahasiswa di kelas paling banyak, yaitu 15, memiliki rata-rata nilai yaitu 78.666667 (dilihat dari standar deviasinya, maka mean tepat dijadikan acuan sebagai nilai tengah data). Yang mana rata-rata nilai tersebut terletak di bawah  CS01 (dengan frekuensi 7 mahasiswa) dan KAL02 (dengan frekuensi 4)

# Code 06 #3

In [ ]:
# Code soal No 06 #3
merge1 = pd.merge(dosen, nilai, on = 'NIDN')
merge1 = merge1[['NIDN', 'nama', 'id_mata_kuliah', 'nilai']]
merge1 = merge1[['nama', 'id_mata_kuliah', 'nilai']]

merge1_mean = merge1.groupby(['nama', 'id_mata_kuliah'])[['nilai']].mean()
merge1_mean = merge1_mean.rename({'nilai': 'mean_nilai'}, axis=1)

merge1_std = merge1.groupby(['nama', 'id_mata_kuliah']).std()[['nilai']]
merge1_std = merge1_std.rename({'nilai': 'std_nilai'}, axis=1)

dosen_mk = pd.concat([merge1_mean, merge1_std], axis=1)
dosen_mk

mean_nilai  std_nilai
nama          id_mata_kuliah                       
Andri Basuki  KAL01            85.750000   7.741355
Annisa Fitria KAL02            63.500000   8.103497
              STA02            58.142857  19.836234
yudi          KAL01            70.571429  13.010984
zulfikar      CS01             80.428571  15.883504

# Interpretasi 06 #3

Terdapat informasi lain dalam data, yaitu berdasarkan data wrangling di atas, akan ditinjau rata-rata pemberian nilai berdasarkan dosen.

* Terdapat 2 dosen yang mengajar mata kuliah yang sama, yaitu Andri Basuki dan Yudi pada mata kuliah KAL01
* Terdapat perbedaan yang signifikan dari nilai yang di berikan kedua dosen
* Andri Basuki cenderung memberi nilai yang lebih baik pada mata kuliah KAL01 daripada Yudi.

Dapat disimpulkan, terdapat dosen yang memberi nilai cenderung lebih baik di banding dosen lain pada mata kuliah yang sama

# Code 06 #4

In [ ]:
# Pemberian nilai Dosen terhadap Mahasiswa berdasarkan Gender
mergea = pd.merge(mahasiswa, nilai, on = 'nim')
mergea = mergea[['NIDN', 'Gender', 'nilai']]

mergeb = pd.merge(mergea, dosen, on = 'NIDN')
mergeb = mergeb[['Jenis_Kelamin', 'Gender', 'nilai']]
mergeb = mergeb.rename(columns={'Jenis_Kelamin': 'Gender_Dosen', 'Gender': 'Gender_Mahasiswa'})

mergeb_mean = mergeb.groupby(['Gender_Dosen', 'Gender_Mahasiswa'])[['nilai']].mean()
mergeb_mean = mergeb_mean.rename({'nilai': 'mean_nilai'}, axis=1)

mergeb_std = mergeb.groupby(['Gender_Dosen', 'Gender_Mahasiswa']).std()[['nilai']]
mergeb_std = mergeb_std.rename({'nilai': 'std_nilai'}, axis=1)

dosen_df = pd.concat([mergeb_mean, mergeb_std], axis=1)
dosen_df

mean_nilai  std_nilai
Gender_Dosen Gender_Mahasiswa                       
Pria         Pria               83.666667  10.368951
             Wanita             73.900000  15.351077
Wanita       Pria               54.333333   8.640988
             Wanita             67.000000  21.330729

# Interpretasi 06 #4

* Dosen dengan gender Pria cenderung memberi nilai yang lebih tinggi terhadap mahasiswa dengan gender yang sama yakni Pria, berlaku demekian untuk dosen dengan gender Wanita. Akan tetapi, masih terdapat variabel lain yang mempengaruhi kriteria seorang dosen dalam memberikan nilai pada seorang mahasiswa.

# Code 06 #5

In [ ]:
mahasiswa_m = mahasiswa[['nim','Asal_Daerah']]
nilai_m = nilai[['nim','nilai']]

int_6 = pd.merge(mahasiswa_m,nilai_m, on = 'nim')
mean = int_6.groupby('Asal_Daerah')[['nilai']].agg('mean')
mean = mean.rename({'nilai': 'mean_nilai'}, axis=1)

std = int_6.groupby(int_6['Asal_Daerah']).std()[['nilai']]
std = std.rename({'nilai': 'std_nilai'}, axis=1)

int_6 = pd.concat([mean,std], axis = 1)

int_6

mean_nilai  std_nilai
Asal_Daerah                       
Lokal         75.764706  17.953166
Luar Daerah   69.750000  15.673757

# Interpretasi 06 #5

Terdapat informasi lain dari dataset, yaitu berdasarkan asal daerah mahasiswanya, mahasiswa asal daerah lokal memiliki mean nilai yang lebih tinggi (mean dipakai sebagai nilai tengah karena standar deviasi cenderung kecil) daripada mahasiswa asal luar daerah.